### PR TITLE
C++ custom result type used for serialization results

### PR DESCRIPTION
**Pull request by Wumpf**
* Depends on https://github.com/rerun-io/rerun/pull/3001
* Part of https://github.com/rerun-io/rerun/issues/2919
* Fixes https://github.com/rerun-io/rerun/issues/2917
* Solves a big chunk of https://github.com/rerun-io/rerun/issues/2873

### What

Introduces a very simple `rerun::Result`. I decided to keep it **a lot** more simple than the arrow Result type and refrained from the typical "return or assign" etc. macro since I noticed that they get quite complicated quickly.

The idea is that the rerun result type won't be needed by a lot of manual code, so erring on the too lightweight side should be in our favor.

We use this now accross the entire public serialization path - meaning that once we move out array->arrow serialization helper into separate headers we should be pretty much done with not exposing arrow headers!

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3005) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3005)
- [Docs preview](https://rerun.io/preview/pr%3Aandreas%2Fcpp%2Fcustom-result-type/docs)
- [Examples preview](https://rerun.io/preview/pr%3Aandreas%2Fcpp%2Fcustom-result-type/examples)

### CODE REVIEW

File: crates/re_types_builder/src/codegen/cpp/mod.rs [hunk 1]
```diff
@@ -1203,8 +1209,11 @@ fn archetype_to_data_cells(
             quote! {
                 if (#field_name.has_value()) {
                     const auto& value  = #field_name.value();
-                    ARROW_ASSIGN_OR_RAISE(const auto cell, #to_data_cell);
-                    cells.push_back(cell);
+                    const auto result = #to_data_cell;
+                    if (!result.ok()) {
+                        return result.status;
+                    }
+                    cells.push_back(result.value);
```
Comment (emilk) [inline crates/re_types_builder/src/codegen/cpp/mod.rs:1216]:
```suggestion
                    cells.push_back(std::move(result.value));
```

here and below?

File: rerun_cpp/src/rerun/arrow.cpp [hunk 2]
```diff
@@ -5,11 +5,17 @@
...
         ARROW_RETURN_NOT_OK(writer->WriteTable(table));
         ARROW_RETURN_NOT_OK(writer->Close());
-        return output->Finish();
+
+        auto result = output->Finish();
+        if (!result.ok()) {
+            return result.status();
+        } else {
+            return result.ValueOrDie();
+        }
```
Comment (emilk) [inline rerun_cpp/src/rerun/arrow.cpp:19]: swap if/else to avoid unnecessary negation

File: rerun_cpp/src/rerun/recording_stream.hpp [hunk 3]
```diff
@@ -180,7 +181,10 @@ namespace rerun {
 
             std::vector<DataCell> data_cells;
             data_cells.reserve(sizeof...(Ts));
-            push_data_cells(data_cells, component_array...);
+            const auto status = push_data_cells(data_cells, component_array...);
+            if (!status) {
```
Comment (emilk) [inline rerun_cpp/src/rerun/recording_stream.hpp:185]: I vote we should be more explicit

```suggestion
            if (status.is_err()) {
```

File: rerun_cpp/src/rerun/recording_stream.hpp [hunk 4]
```diff
@@ -216,36 +220,44 @@ namespace rerun {
...
-        static void push_data_cells(
+        static Status push_data_cells(
             std::vector<DataCell>& data_cells, const std::vector<C>& first, const Ts&... rest
         ) {
-            // TODO(andreas): Error handling.
-            const auto cell = C::to_data_cell(first.data(), first.size()).ValueOrDie();
-            data_cells.push_back(cell);
-            push_data_cells(data_cells, rest...);
+            const auto cell_result = C::to_data_cell(first.data(), first.size());
+            if (!cell_result.ok()) {
```
Comment (emilk) [inline rerun_cpp/src/rerun/recording_stream.hpp:227]: we can follow the Rust name convention? 🤷 

```suggestion
            if (cell_result.is_err()) {
```

File: rerun_cpp/src/rerun/result.hpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        Result(arrow::Status&& error) : value(), status(std::move(error)) {}
+
+        /// Returns true if status was ok and a value is contained, false otherwise.
+        bool ok() const {
+            return status;
+        }
+
+#ifdef __cpp_exceptions
+        /// Returns the value if status is ok, throws otherwise.
+        T value_or_throw() {
```
Comment (emilk) [inline rerun_cpp/src/rerun/result.hpp:43]:
```suggestion
        T value_or_throw() const {
```

File: rerun_cpp/src/rerun/status.cpp [hunk 6]
```diff
@@ -9,6 +10,66 @@ namespace rerun {
...
+                code = StatusCode::ArrowStatusCode_ExpressionValidationError;
+                break;
+            case arrow::StatusCode::ExecutionError:
+                code = StatusCode::ArrowStatusCode_ExecutionError;
+                break;
+            case arrow::StatusCode::AlreadyExists:
+                code = StatusCode::ArrowStatusCode_AlreadyExists;
+                break;
+            default:
+                code = StatusCode::Unknown;
```
Comment (emilk) [inline rerun_cpp/src/rerun/status.cpp:67]:
```suggestion
                code = StatusCode:: ArrowStatusCode_Unknown;
```

File: rerun_cpp/src/rerun/status.hpp [hunk 7]
```diff
@@ -9,23 +9,28 @@
...
     /// Status codes returned by the SDK as part of `Status`.
     ///
     /// Category codes are used to group errors together, but are never returned directly.
     enum class StatusCode : uint32_t {
         Ok = 0,
+        OutOfMemory = 1,
 
         // Invalid argument errors.
-        _CategoryArgument = 0x000000010,
+        _CategoryArgument = 0x00000010,
```
Comment (emilk) [inline rerun_cpp/src/rerun/status.hpp:25]: this can help readability:

```suggestion
        _CategoryArgument = 0x0000'0010,
```

**Review by emilk — APPROVED** [review]
Nice!

File: rerun_cpp/src/rerun/status.hpp (hunk 7 above)
Comment (Wumpf) [reply rerun_cpp/src/rerun/status.hpp:25]: oh nice didn't know that's a thing. I tried 0x0000_000 and it didn't like that

File: rerun_cpp/src/rerun/status.cpp (hunk 6 above)
Comment (Wumpf) [reply rerun_cpp/src/rerun/status.cpp:67]: I'd argue that's different - we don't know if it's an error unknown, instead it is an error that we literally don't know about

**Pull request merged**